### PR TITLE
Rewrite the Orleans serialization content with `7.0` in mind

### LINKED ISSUE #32736: Rewrite serialization documentation for Orleans 7.0
**Issue description**

Orleans 7.0 replaced the pre-existing serialization framework and the docs now need to be rewritten from scratch to match.
For more information, see https://learn.microsoft.com/en-us/dotnet/orleans/whats-new-in-orleans#serialization

The linked section should serve as a basis for the new docs, but it will need to be rewritten to use the right voice, etc.

Ideally, we need to keep the existing 3.x documentation around so that people can continue to refer to it. I don't know how to do that and will need help.

cc @IEvangelist @bradygaster 

---
[Associated WorkItem - 46074](https://dev.azure.com/msft-skilling/Content/_workitems/edit/46074)

### PR DESCRIPTION
**Pull request by IEvangelist**
## Summary

In this PR:

- Split content into multiple articles, Overview, Customization, Configuration, and Immutable types.
- Add zone pivots for `3.x` and `7.0`.
- Add new content for introduction paragraphs, custom serialization impls.

Fixes #32736 /cc @ReubenBond @bradygaster 

Update Serialization Content Previews:

- [Serialization overview](https://review.learn.microsoft.com/en-us/dotnet/orleans/host/configuration-guide/serialization?branch=pr-en-us-32990)
- [Serialization of immutable types](https://review.learn.microsoft.com/en-us/dotnet/orleans/host/configuration-guide/serialization-immutability?branch=pr-en-us-32990)
- [Serialization configuration](https://review.learn.microsoft.com/en-us/dotnet/orleans/host/configuration-guide/serialization-configuration?branch=pr-en-us-32990&pivots=orleans-7-0)
- [Serialization customization](https://review.learn.microsoft.com/en-us/dotnet/orleans/host/configuration-guide/serialization-customization?branch=pr-en-us-32990&pivots=orleans-7-0)

Updated TOC:

image

### CODE REVIEW

File: docs/orleans/host/configuration-guide/serialization-configuration.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+<!-- markdownlint-disable MD044 -->
+:::zone target="docs" pivot="orleans-7-0"
+<!-- markdownlint-enable MD044 -->
+
+The configuration of serialization in Orleans is a crucial part of the overall system design. While Orleans provides reasonalble defaults, you can configure serialization to suit your apps' needs. For sending data between hosts, <xref:Orleans.Serialization?displayProperty=fullName> supports delegating to other serializers, such as [Newtonsoft.Json](https://www.nuget.org/packages/Newtonsoft.Json) and [System.Text.Json](https://www.nuget.org/packages/System.Text.Json). You can add support for other serializers by following the pattern set by those implementations. For grain storage it is preferential to use <xref:Orleans.Storage.IGrainStorageSerializer> to configure a custom serializer.
+
+## Configure Orleans to use `System.Text.Json`
+
+Configuring Orleans to use `System.Text.Json` to serialize a subset of types is similar to configuring Orleans to serialize types using `Newtonsoft.Json`, except that the package and configuration methods are different:
```
Comment (ReubenBond) [inline docs/orleans/host/configuration-guide/serialization-configuration.md:19]: This forward-references the Newtonsoft.Json section, but they won't have read that yet

**Review by tdykstra — APPROVED** [review]
This is very well written! Clear and easy to understand. My suggestions are mostly nits.

File: docs/orleans/host/configuration-guide/serialization-customization.md [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+            SerializationManager.DeserializeInner<DateTimeOffset>(context);
+
+        return result;
+    }
+}
+```
+
+### Serialize generic types
+
+The `TargetType` parameter of `[Serializer(typeof(TargetType))]` can be an open-generic type, for example, `MyGenericType<>`. In that case, the serializer class must have the same generic parameters as the target type. Orleans will create a concrete version of the serializer at runtime for every concrete `MyGenericType<T>` type which is serialized, for example, one for each of `MyGenericType<int>` and `MyGenericType<string>`.
```
[the remark this answers is not in the document]
Comment (IEvangelist) [reply docs/orleans/host/configuration-guide/serialization-customization.md:378]: Yes, good catch.

File: docs/orleans/host/configuration-guide/serialization.md [hunk 3]
```diff
@@ -1,353 +1,282 @@
...
 
-Serializers are registered for each supported data type at silo start-up and whenever an assembly is loaded. Registration is necessary for custom serializer routines for a type to be used. Serializer selection is based on the dynamic type of the object to be copied or serialized. For this reason, there is no need to create serializers for abstract classes or interfaces, because they will never be used.
+:::image type="content" source="media/generate-serializer-code-fix.gif" alt-text="An animated image of the available code fix being suggested and applied on the GenerateSerializerAttribute when the containing type doesn't contain IdAttribute's on its members." lightbox="media/generate-serializer-code-fix.gif":::
```
[the remark this answers is not in the document]
Comment (IEvangelist) [reply docs/orleans/host/configuration-guide/serialization.md:31]: That's a good question, I'll look into that.

[the remark this answers is not in the document]
Comment (IEvangelist) [reply docs/orleans/host/configuration-guide/serialization.md:31]: Slow it down.

File: docs/orleans/host/configuration-guide/serialization-configuration.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+description: Learn how to configure serialization in .NET Orleans.
+ms.date: 12/15/2022
+uid: orleans-serialization-configuration
+zone_pivot_groups: orleans-version
+---
+
+# Serialization configuration in Orleans
+
+<!-- markdownlint-disable MD044 -->
+:::zone target="docs" pivot="orleans-7-0"
```
Comment (bradygaster) [inline docs/orleans/host/configuration-guide/serialization-configuration.md:12]: 💯

**Pull request merged**